### ISSUE TITLE
Resnet - converted Onnx model is 2.9X slower than pyTorch model in V100 gpu

### ISSUE THREAD

**Issue by yetingqiaqia** [microsoft/onnxruntime#9162]
**Describe the bug**
Hi ort team,

We are using ORT to serve our models. 
Our original models are torch models. We will first convert torch models to ONNX models using torch.onnx.export(opset_version=x) and run inference with the converted ONNX models. We do it for speed, usually, ONNX model can be 1.3x~2x faster than original pyTorch model.

However, recently, we met a resnet model. To our surprise, after converted to onnx model, its speed is 2.9x slower than original pyTorch model.  We would like to ask your help to figure out why and how to resolve it. Thanks.

Below is the test result:
- The number of input samples is 15k

| Framework | Inference Time (s) | Throughput(samples/s/gpu) |
| ------------- | ------------- | ------------- |
| PyTorch | 248.95 | **60.25** |
| Onnx+Opset12 | 721.74 | **20.78** |
| Onnx+Opset13 | 725.58 | **20.67**  |

Note:
1. All the tests are based on the same settings, both models are using fp32, with the same batch_size=32. 
2. We tested several torch versions (1.9.1, 1.8.0), onnxruntime versions (1.8.1, 1.6.0), Onnx opset version (opset12 and opset13) and GPU types (V100 and A100), they had the same behaviors. To keep this ticket clear and easy to reproduce, the numbers I reported in the table were run in ORT1.8.1, torch1.9.1+cu111 in V100, all packages are the latest versions. But even test on other torch versions, onnxruntime versions or GPU types, this issue is consistent. 

**Urgency**
We are running an ONNX inference platform for Microsoft ads users. This is a blocking issue for users running such kinds of models.

**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04.2 LTS
- ONNX Runtime installed from (source or binary): binary
- ONNX Runtime version: 1.8.1, install command: pip install onnxruntime-gpu==1.8.1
- Torch version: 1.9.1+cu111, install command: pip3 install torch==1.9.1+cu111 torchvision==0.10.1+cu111 torchaudio==0.9.1 -f https://download.pytorch.org/whl/torch_stable.html
- Python version: 3.6.8
- Visual Studio version (if applicable): N/A
- GCC/Compiler version (if compiling from source): N/A
- CUDA/cuDNN version: Driver Version: 455.45.01    CUDA Version: 11.1 
- GPU model and memory: V100, 16GB
 We also tested it in (A100, 40GB.  AML SKU: Standard_ND96asr_v4 ) with batch_size=32 and 128. The results are consistent. 

**To Reproduce**
- Describe steps/code to reproduce the behavior.

Below is the core code snippet that I used to covert torch model to onnx model and test their inference time. Detailed code, model and data are attached below.
* convert_torch_to_onnx(): convert torch model to onnx model
* torch_evaluate(): torch inference
* onnx_evaluate(): onnx inference

```python
def load_torch_model(args):
    model_name="BiT-M-R152x2"
    model = models.KNOWN_MODELS[model_name](head_size=3, zero_head=True)
    model = torch.nn.DataParallel(model)
    try:
        checkpoint = torch.load(args.torch_model_path, map_location="cpu")
        print(f"Found saved model to resume from at '{args.torch_model_path}'")
        step = checkpoint["step"]
        model.load_state_dict(checkpoint["model"])
        print(f"Resumed at step {step}", )
    except FileNotFoundError:
        print("Fine-tuning from BiT", flush=True)
    model.to(args.device)
    model.eval()
    return model

def convert_torch_to_onnx(args):
    torch_model = load_torch_model(args)
    batch_size = 1 # batch_size is only used to generate the random input data.
    image_size = 480 # max_seq_len is corresponding to training, in our example roberta model, the max_seq_len is 32
    input_image       = torch.FloatTensor(torch.zeros((batch_size, 3, image_size, image_size), dtype=torch.float))
    input_image = input_image.cuda()
    input_tensors =  (input_image, )
    input_names = ["input_image"]
    output_names = ["logits"]
    dynamic_axes={'input_image':{0:'batch_size'}, 'logits':{0:'batch_size'}}
    print("model export begins ... ")
    torch.onnx.export(
        torch_model.module,     # model being run
        input_tensors,          # model input (or a tuple for multiple inputs)
        args.output_onnx_model_path,   # where to save the model (can be a file or file-like object)
        export_params=True,        # store the trained parameter weights inside the model file
        opset_version=args.opset_version,          # the ONNX version to export the model to
        do_constant_folding=True,  # whether to execute constant folding for optimization
        verbose=True,              # verbose=True causes the exporter to print out a human-readable representation of the network
        input_names = input_names,   # the model's input names
        output_names = output_names, # the model's output names
        dynamic_axes= dynamic_axes,  # variable lenght axes
        use_external_data_format=False # should be true for model size >2GB
    )
    print("model export done.")

def torch_evaluate(args):
    torch_model = load_torch_model(args)
    val_loader = get_val_loader(args)

    accumulated_inference_time = 0
    with torch.no_grad():
        for i, (id, data, target) in enumerate(val_loader):
            data = data.to(args.device)
            #below is pytorch inference code
            start_stamp = time.time()
            #pred = torch_model(data)
            with torch.cuda.amp.autocast():
                pred = torch_model(data)
            torch.cuda.synchronize()
            accumulated_inference_time += time.time() - start_stamp

    print(f"Total Pytorch model inference time is {accumulated_inference_time}")

def onnx_evaluate(args):
    val_loader = get_val_loader(args)
    import onnxruntime as ort
    sess_options = ort.SessionOptions()
    sess_options.graph_optimization_level = ort.GraphOptimizationLevel.ORT_ENABLE_ALL
    sess_options.intra_op_num_threads = 0
    sess = ort.InferenceSession(args.onnx_model_path, sess_options)
    input_name = sess.get_inputs()[0].name
    label_name = sess.get_outputs()[0].name
    def to_numpy(torch_tensor):
        return torch_tensor.detach().cpu().numpy() if torch_tensor.requires_grad else torch_tensor.cpu().numpy()

    accumulated_inference_time = 0
    with torch.no_grad():
        for i, (id, data, target) in enumerate(val_loader):
            data = data.to(args.device)
            #below is onnx inference code
            start_stamp = time.time()
            pred = sess.run([label_name], {input_name: to_numpy(data)})[0]
            accumulated_inference_time += time.time() - start_stamp

    print(f"Total Onnx model inference time is {accumulated_inference_time}")

if __name__ == "__main__":
    script_start_time = time.time()
    parser = argparse.ArgumentParser()
    parser.add_argument("--mode", type=str, help='torchInference, onnxInference, or torch2Onnx')

    args, unknown = parser.parse_known_args()

    args.batch_size = 32
    args.image_size = 480
    args.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    args.n_gpu = torch.cuda.device_count()
    print(" ARGS.device " + str(args.device) + "  ARGS.n_gpu " + str(args.n_gpu))

    args.test_file_path = "./data/input.tsv"
    if args.mode == "torchInference":
        args.torch_model_path = "./model/bit.pth.tar"
        args.outputfilename = './output/torch_result.tsv'
        torch_evaluate(args)
    elif args.mode == "onnxInference":
        args.onnx_model_path = "./model/graph_opset12.onnx"
        args.outputfilename = './output/onnx_opset12_result.tsv'
        onnx_evaluate(args)
    elif args.mode == "torch2Onnx":
        args.torch_model_path = "./model/bit.pth.tar"
        args.output_onnx_model_path = "./model/graph_opset13.onnx"
        args.opset_version = 13
        convert_torch_to_onnx(args)
    else:
        raise Exception("Unrecognized --mode, please use torchInference, onnxInference, or torch2Onnx")
```

- Attach the ONNX model to the issue (where applicable) to expedite investigation.
https://drive.google.com/file/d/1EVqAsYAaju9vU5AdOlmAYXRWAvcZHIG3/view?usp=sharing

It contains 6 files:
1. resnetInference_torch_vs_onnx.py: test code
2. models.py: model definition code
3. ./data/input.tsv: input data
4. ./model/bit.pth.tar: original torch model
5. ./model/graph_opset12.onnx: converted onnx opset12
6. ./model/graph_opset13.onnx: converted onnx opset13

Commands:
1. To get the inference time of Torch model: ```python resnetInference_torch_vs_onnx.py --mode torchInference```
2. To get the inference time of Onnx model: ```python resnetInference_torch_vs_onnx.py --mode onnxInference```
3. To convert Torch model to onnx model: ```python resnetInference_torch_vs_onnx.py --mode torch2Onnx```

**Expected behavior**
I expect the inference speed of the converted onnx model should be faster than the original torch model, at least not slower than torch model.

**Comment by hariharans29** [comment]
I wonder if one or more CUDA kernels are not performant enough for the batched input use case (which we can identify and fix by profiling). Just curious, (if it is possible) can you benchmark with batch size = 1 ? If batching is the problem for some kernels, the performance when there is only a single input should be better. FWIW - Can you try with the [latest nightly](https://test.pypi.org/project/ort-nightly-gpu/) to see if the performance is better ?

**Comment by yetingqiaqia** [comment]
Thanks @hariharans29 . I tested batch_size=1 with 2k samples. 
Here is result:

- PyTorch: 115.77s, i.e., 17.28 samples/s/gpu
- Onnx+opset13: 190.23s, i.e., 10.51 samples/s/gpu

The gap is slightly narrowed, but it is still 1.67X slower. 

For latest nightly version, I will have a try and report it later.

**Comment by hariharans29** [comment]
Thanks.

For the Torch evaluation, you are copying the data to the device and then starting the timer:

'''
            data = data.to(args.device)
            #below is pytorch inference code
            start_stamp = time.time()
'''

The result from Torch stays on CUDA (I think!)

While evaluating ORT, the data is copied over to CPU and then taken to the device (as part of Run()) after the timer is turned on and the measured time also includes result data transfer from GPU to CPU

To truly get an apple-to-apple comparison with Torch, can you modify the benchmarking script to use [IOBinding](https://github.com/microsoft/onnxruntime/blob/6e83392ff134f450aa5b8798ccbb138ec32b5379/onnxruntime/test/python/onnxruntime_test_python_iobinding.py#L97) and turn on the timer after the data has been copied over to the device (just like the Torch evaluation part) ?

**Comment by yetingqiaqia** [comment]
@hariharans29 , for latest nightly version (v1.9.0). It performs quite bad. I can't use Batch_Size=32 any more because of "out of shared memory" issue. So, I switch to BS=16 to test. The inference time on 15k is 2333.4 s, which is about 3.2X worse than using official 1.8.1 version.
- The command I used to install latest nightly version: pip install -i https://test.pypi.org/simple/ ort-nightly-gpu

For the input data part, according to my previous experience, it doesn't matter much, especially for heavy models like this. I don't think it could reduce the 65% (the time gap between pytorch and ORT on this model) by changing IO.

**Comment by hariharans29** [comment]
Thanks.

1) You're right - usually IOBinding doesn't matter too much if the batch size is 1. But here since batch size is >> 1, I would think it makes a bigger difference than usual (as the input buffer and output buffer is correspondingly large). It is just 2-3 lines of code change. Please consider doing it to get the "real" gap.

 2) For Conv heavy models, it is important to do a warm-up prior to benchmarking. The very first run is very expensive compared to the other runs because the best Conv algo is established only on the first run. So, please add a warm-up run with the same input shape as the rest of the runs. This is to  establish the "true" gap. 

3) The regression in the nightly (1.9's performance should be the same as the nightly) compared to 1.8 is to be investigated as is the original poor perf in 1.8 (with warm-up and IOBinding, the true gap for this will be known).

**Comment by yetingqiaqia** [comment]
@hariharans29 for IOBinding, I tried to follow the example script you shared and tested with top 10 samples.
However, I tested two ways and both met some issues:
1. Using a static input shape [1, 3, 480, 480]. 
            ```io_binding.bind_input(input_name, 'cuda', 0, np.float32, [1, 3, 480, 480], data.data_ptr())```
     This one is runnable, but it will only process the first sample of each batch:
      - Expected output: 
![image](https://user-images.githubusercontent.com/6299908/134750278-1da95a5e-72df-4ea9-aae4-898067b27e47.png)
      - Real output:
![image](https://user-images.githubusercontent.com/6299908/134750296-46cc1b23-06ef-47d9-afbe-20595d0470c1.png)
       
      As you can see, by this way, it only processes the first sample.
2. To avoid issue in case1, I calculated the data_shape instead of using a static shape
            ```data_shape = list(to_numpy(data).shape)```
            ```io_binding.bind_input(input_name, 'cuda', 0, np.float32, data_shape, data.data_ptr())```
      However, this way, the script failed because of "OrtValue shape verification failed. Current shape:{1,3} Requested shape:{2,3}" error.
![image](https://user-images.githubusercontent.com/6299908/134750525-14db6b1b-62b5-411a-aa51-0cbafe3f81a0.png)

I am not familiar with ORT IOBinding. **Coud you help check what's wrong and how to make the code runnable?**
Also, for the output ```pred = io_binding.get_outputs()[0].numpy()```, **what is the meaning of "[0]" here? Does it mean the output of the first output tensor? If we have 2 tensors, then there will be [1] for the second output tensor?** 

Below is my modified code:
```python
def onnx_evaluate(args):
    val_loader = get_val_loader(args)
    import onnxruntime as ort
    sess_options = ort.SessionOptions()
    sess_options.graph_optimization_level = ort.GraphOptimizationLevel.ORT_ENABLE_ALL
    sess_options.intra_op_num_threads = 0
    sess = ort.InferenceSession(args.onnx_model_path, sess_options)
    input_name = sess.get_inputs()[0].name
    label_name = sess.get_outputs()[0].name

    io_binding = sess.io_binding()
    io_binding.bind_output(label_name, 'cuda')
    def to_numpy(torch_tensor):
        return torch_tensor.detach().cpu().numpy() if torch_tensor.requires_grad else torch_tensor.cpu().numpy()

    accumulated_inference_time = 0

    f = open(args.outputfilename, 'w', encoding='utf-8')
    f.write('\t'.join(['Id', 'Score']) + '\n')

    with torch.no_grad():
        for i, (id, data, target) in enumerate(val_loader):
            data = data.to(args.device)
            data_shape = list(to_numpy(data).shape)
            print(data_shape)
            #io_binding.bind_input(input_name, 'cuda', 0, np.float32, [1, 3, 480, 480], data.data_ptr())
            io_binding.bind_input(input_name, 'cuda', 0, np.float32, data_shape, data.data_ptr())
            #below is onnx inference code
            start_stamp = time.time()
            sess.run_with_iobinding(io_binding)
            #pred = sess.run([label_name], {input_name: to_numpy(data)})[0]
            accumulated_inference_time += time.time() - start_stamp
            pred = io_binding.get_outputs()[0].numpy()
            f.write('\t'.join([str(id),
                               ','.join([str(l) for l in pred])]) + '\n')
    print(f"Total Onnx model inference time is {accumulated_inference_time}")

    print(f"The file is present at location: {args.outputfilename}")
    f.flush()
    f.close()
```

**Comment by hariharans29** [comment]
1) This is expected. The data being bound is just the first sample because the input shape of bound data is [**1**, 3, 480, 480] and so processing is only being done for that sample of the batch. Your second approach is correct.

2) You are seeing this because the input shape is changing (batch size changes from 1 to 2) and the output is not re-bound. Please add this line  `io_binding.bind_output(label_name, 'cuda')` after bind_input() and before starting the timer. Ideally, the input shape will remain the same across all iterations of the benchmark if that is your prod scenario. Please run a warm-up run for that input shape to cache some CuDNN details prior to bench-marking.

**Comment by hariharans29** [comment]
Closing this as all the issues pertaining to this model were resolved through : https://github.com/microsoft/onnxruntime/pull/9826, https://github.com/microsoft/onnxruntime/pull/9879, using IOBinding (see script below), and allowing CuDNN to use maximum workspace memory during Conv algo picking (see script below).

Final inferencing script for reference:

'''
import numpy as np
import time
import onnxruntime as ort

# Batch size  = 8
input = np.random.rand(8, 3, 480, 480).astype(np.float32)
num_batches = 200

sess_options = ort.SessionOptions()
sess = ort.InferenceSession("resnet_fp16.onnx", sess_options, providers=['CUDAExecutionProvider'])
input_name = sess.get_inputs()[0].name
label_name = sess.get_outputs()[0].name


options = sess.get_provider_options()
cuda_options = options['CUDAExecutionProvider']
cuda_options['cudnn_conv_use_max_workspace'] = '1'
sess.set_providers(['CUDAExecutionProvider'], [cuda_options])

io_binding = sess.io_binding()
io_binding.bind_output(label_name, 'cuda')
data = ort.OrtValue.ortvalue_from_numpy(input, 'cuda', 0)
io_binding.bind_input(input_name, 'cuda', 0, np.float32, [8, 3, 480, 480], data.data_ptr())

#warm-up run (dis-regard latency)
warm_up_start_stamp = time.time()
sess.run_with_iobinding(io_binding)
pred = io_binding.copy_outputs_to_cpu()[0]
print(f"It takes {time.time()-warm_up_start_stamp} to finish warm-up.\n"
+f" - cudnn_conv_use_max_workspace is {sess.get_provider_options()['CUDAExecutionProvider']['cudnn_conv_use_max_workspace']}\n"
+f" - io_binding")
#print(pred)

# Perf benchmark
accumulated_inference_time = 0
for i in range(num_batches):
#data = ort.OrtValue.ortvalue_from_numpy(input, 'cuda', 0)
#io_binding.bind_input(input_name, 'cuda', 0, np.float32, [8, 3, 480, 480], data.data_ptr())
#below is onnx inference code
start_stamp = time.time()
sess.run_with_iobinding(io_binding)
#pred = sess.run([label_name], {input_name: to_numpy(data)})[0]
accumulated_inference_time += time.time() - start_stamp
pred = io_binding.copy_outputs_to_cpu()[0]
print(pred)

print(f"It takes {accumulated_inference_time} to finish {num_batches} batches.\n"
+f" - cudnn_conv_use_max_workspace is {sess.get_provider_options()['CUDAExecutionProvider']['cudnn_conv_use_max_workspace']}\n"
+f" - io_binding")
'''

**Comment by yetingqiaqia** [comment]
Thanks @hariharans29 for driving this. It is good to see ORT is faster than torch on this model finally!
